File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, int> ump;
    int totalNodes(TreeNode* root){
        if(!root) return 0;
        if(ump.find(root)!=ump.end()) return ump[root];
        return ump[root]= 1+totalNodes(root->left)+totalNodes(root->right);
    }
    int kthSmallest(TreeNode* root, int k) {
        int lft= totalNodes(root->left);
        if(lft>=k) return kthSmallest(root->left, k);
        k-=lft;
        if(k==1) return root->val;
        return kthSmallest(root->right, k-1);
    }
};

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int cnt;
    TreeNode ans;
    private void inOrder(TreeNode root, int k){
        if(root==null) return;
        inOrder(root.left, k);
        cnt++;
        if(cnt==k) ans= root;
         inOrder(root.right, k);
    }
    public int kthSmallest(TreeNode root, int k) {
        cnt=0; ans=null;
        inOrder(root, k);
        return ans.val;
    }
}